### PR TITLE
Enable cuik-molmaker for accelerated molecule featurization

### PR DESCRIPTION
**Pull request by sveccham**
Enable cuik-molmaker for accelerated molecule featurization

## Description
This PR enables the use of cuik-molmaker (https://github.com/NVIDIA-Digital-Bio/cuik-molmaker), a C++/python package that accelerates atom and bond featurization. Usage of cuik-molmaker accelerates Chemprop training by 1.6X and inference by 2.4X. In addition, memory usage is reduced by ~80% enables larger-scale training and inference workloads. 

- Supports `train.py`, `predict.py`, `fingerprint.py`, and `hpopt.py`
- Support multiple and and bond featurization modes: `V1`, `V2`, `ORGANIC`, `RIGR`
- Currently, cuik-molmaker works with conda installs ([option 3](https://chemprop.readthedocs.io/en/latest/installation.html#option-3-installing-from-source-using-environment-yml)) and Docker container ([option 4](https://chemprop.readthedocs.io/en/latest/installation.html#option-4-installing-via-docker))
- Added download script `chemprop/scripts/check_and_install_cuik_molmaker.py` that simplifies the installation process of cuik-molmaker
- Added unit tests

## Example / Current workflow
`CUDA_VISIBLE_DEVICES=0 python  chemprop/cli/train.py --data-path path/to/data.csv --smiles-columns smiles --target-columns logSolubility --task-type regression --epochs 5 --data-seed 0 --pytorch-seed 0 --multi-hot-atom-featurizer-mode V1 --no-cache`

## Bugfix / Desired workflow
`CUDA_VISIBLE_DEVICES=0 python  chemprop/cli/train.py --data-path path/to/data.csv --smiles-columns smiles --target-columns logSolubility --task-type regression --epochs 5 --data-seed 0 --pytorch-seed 0 --multi-hot-atom-featurizer-mode V1 --no-cache --use-cuikmolmaker-featurization`

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by sveccham** [comment]
@KnathanM , looks like the full test suite passes on my end. Could you please trigger the CI workflow?

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -979,9 +979,11 @@ def build_splits(args, format_kwargs, featurization_kwargs):
 
     else:
         splitting_data = all_data[args.split_key_molecule]
+
         if isinstance(splitting_data[0], ReactionDatapoint):
             splitting_mols = [datapoint.rct for datapoint in splitting_data]
         else:
+            # TODO: Form mols only if needed by the splitting type; Random split does not need mols
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:986]: Could you implement this, or make an issue to track it?

File: chemprop/cli/utils/MAB_parsing.py [hunk 2]
```diff
@@ -212,7 +212,15 @@ def build_MAB_data_from_files(
...
-            mol=mols[i],
+            smiles=smis[i],
+            _keep_h=make_mol_kwargs["keep_h"] if "keep_h" in make_mol_kwargs else False,
+            _add_h=make_mol_kwargs["add_h"] if "add_h" in make_mol_kwargs else False,
+            _ignore_stereo=make_mol_kwargs["ignore_stereo"]
+            if "ignore_stereo" in make_mol_kwargs
+            else False,
+            _reorder_atoms=make_mol_kwargs["reorder_atoms"]
+            if "reorder_atoms" in make_mol_kwargs
+            else False,
```
Comment (JacksonBurns) [inline chemprop/cli/utils/MAB_parsing.py:223]: All of these attempted key accesses will result in a KeyError if they key is not present because the if else is executed greedily (I think) so we need different approach (which is also conveniently a bit more concise):

```suggestion
            _keep_h=make_mol_kwargs.get("keep_h", False),
            _add_h=make_mol_kwargs.get("add_h", False),
            _ignore_stereo=make_mol_kwargs.get("ignore_stereo", False),
            _reorder_atoms=make_mol_kwargs.get("reorder_atoms", False),
```

File: chemprop/cli/utils/parsing.py [hunk 3]
```diff
@@ -214,11 +220,69 @@ def make_datapoints(
...
+                _keep_h=keep_h,
+                _add_h=add_h,
+                _ignore_stereo=ignore_stereo,
+                _reorder_atoms=reorder_atoms,
+                name=smis[i],
+                y=Y[i],
+                weight=weights[i],
+                gt_mask=gt_mask[i],
+                lt_mask=lt_mask[i],
+                # x_d=X_d[i],
```
Comment (JacksonBurns) [inline chemprop/cli/utils/parsing.py:246]: Should not be commented out, perhaps?

File: chemprop/data/datasets.py [hunk 4]
```diff
@@ -333,6 +338,66 @@ def reset(self):
...
+        if cache:
+            warnings.warn(
+                "This option no longer does anything and is deprecated. --use-cuikmolmaker-featurization is meant to be used without caching!",
+                DeprecationWarning,
+            )
+        self.__cache = cache
+
+    def __getitems__(self, indexes: list[int]):
+        # TODO(sveccham): Remove in final PR
+        print("__getitems__ called")
```
Comment (JacksonBurns) [inline chemprop/data/datasets.py:379]: Just a reminder to take this out

File: chemprop/data/datasets.py [hunk 5]
```diff
@@ -333,6 +338,66 @@ def reset(self):
...
+        return self.__cache
+
+    @cache.setter
+    def cache(self, cache: bool = False):
+        if cache:
+            warnings.warn(
+                "This option no longer does anything and is deprecated. --use-cuikmolmaker-featurization is meant to be used without caching!",
+                DeprecationWarning,
+            )
+        self.__cache = cache
```
Comment (JacksonBurns) [inline chemprop/data/datasets.py:375]: We use `logger` for things like this, see e.g.: https://github.com/chemprop/chemprop/blob/c4602da66bf9dc766ea48af126fc2600b445e923/chemprop/nn/utils.py#L35

File: chemprop/scripts/check_and_install_cuik_molmaker.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+conda_prefix = os.getenv("CONDA_PREFIX")
+
+rdkit_so_file = os.path.join(conda_prefix, "lib", "libRDKitGraphMol.so")
+print(f"RDKit shared object file: {rdkit_so_file}")
+if not os.path.exists(rdkit_so_file):
+    print(
+        "RDKit is not installed. Please install RDKit from conda and try again."
+        "You can install RDKit with the following command: conda install -c conda-forge rdkit"
+    )
+    exit(1)
```
Comment (JacksonBurns) [inline chemprop/scripts/check_and_install_cuik_molmaker.py:25]: RDKit should be installed if chemprop is installed, so I'd suggest adding:

```suggestion
rdkit_so_file = os.path.join(conda_prefix, "lib", "libRDKitGraphMol.so")
print(f"RDKit shared object file: {rdkit_so_file}")
if not os.path.exists(rdkit_so_file):
    print(
        f"Could not find RDKit SO file {rdkit_so_file}"
        "This probably means RDKit and/or Chemprop is not installed. Please install RDKit/Chemprop from conda and try again."
        "You can install these with the following command: conda install 'conda-forge::rdkit' 'conda-forge::chemprop'"
    )
    exit(1)
```

Will this work with `pip`? I believe `pip` is smart enough these days to install things into the `conda` `lib` directory, though maybe not.

File: chemprop/scripts/check_and_install_cuik_molmaker.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+libtorch_so_file = os.path.join(conda_prefix, "lib", "libtorch_cpu.so")
+if not os.path.exists(libtorch_so_file):
+    print(
+        "LibTorch is not installed. Please install PyTorch from conda and try again."
+        "You can install LibTorch with the following command: conda install -c conda-forge pytorch"
+    )
+    exit(1)
```
Comment (JacksonBurns) [inline chemprop/scripts/check_and_install_cuik_molmaker.py:33]:
```suggestion
if not os.path.exists(libtorch_so_file):
    print(
        f"Could not fine LibTorch SO file {libtorch_so_file} "
        "This probably means LibTorch is not installed. Please install PyTorch from conda and try again."
        "You can install LibTorch with the following command: conda install -c conda-forge pytorch"
    )
    exit(1)
```

File: chemprop/scripts/check_and_install_cuik_molmaker.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+# Check if rdkit_so_file was compiled with CXX11 ABI by looking for mangled symbols
+try:
+    # Run strings command and pipe through grep to check for cxx11 symbols
+    strings_output = subprocess.check_output(["strings", rdkit_so_file], text=True)
```
Comment (JacksonBurns) [inline chemprop/scripts/check_and_install_cuik_molmaker.py:38]: `strings` will probably not be available on Windows systems. This would lead to a CalledProcessError and a possible negative with this check.

Maybe best to check if the user is running on Windows right from the get-go and `exit(1)` if so?

File: chemprop/scripts/check_and_install_cuik_molmaker.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+except subprocess.CalledProcessError as e:
+    print(f"Failed to install cuik-molmaker: {e}")
+    print(
+        "1. Install from source. Follow instructions at https://github.com/NVIDIA-Digital-Bio/cuik-molmaker"
+    )
+    print("2. Reach out to cuik-molmaker developers at cuik_molmaker_dev@nvidia.com")
+    exit(1)
+
+print("cuik-molmaker installed successfully.")
+print("You can now use cuik-molmaker with Chemprop using --use_cuikmolmaker_featurization flag.")
```
Comment (JacksonBurns) [inline chemprop/scripts/check_and_install_cuik_molmaker.py:156]:
```suggestion
print("You can now use cuik-molmaker with Chemprop using --use-cuikmolmaker-featurization flag.")
```

File: chemprop/utils/utils.py [hunk 10]
```diff
@@ -99,3 +101,25 @@ def pretty_shape(shape: Iterable[int]) -> str:
...
+    # Get the current process
+    process = psutil.Process(os.getpid())
+
+    # Get memory info in bytes
+    memory_info = process.memory_info()
+
+    # Convert to MB for readability
+    memory_mb = memory_info.rss / 1024 / 1024
+
+    return f"Memory usage: {memory_mb:.2f} MB"
```
Comment (JacksonBurns) [inline chemprop/utils/utils.py:116]: Can probably delete this? Or is it used somewhere? Fine to keep it, if so

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
Thanks for you and your team's outstanding work on this PR! We're very excited to see chemprop scale up larger and larger!

A couple overall notes:
 - Rather than changing the behavior of `_MoleculeDatapointMixin` to be Lazy by default, I think a `_LazyMoleculeDatapointMixin` would be better - keeps existing code working exactly the same way.
 - The names `BatchMoleculeDataset` and `BatchMolGraphFeaturizer` make sense but obscure the fact that they use `cuik-molmaker`. For the sake of clarity, I think they should be renamed something along the lines of `CuikMolMakerDataset` and `CuikMolmakerFeaturizer` or something in that direction
  
 I've not yet looked at the notebook or tests - will review those after this initial round (or let another reviewer take a look, perhaps).

File: chemprop/scripts/check_and_install_cuik_molmaker.py (hunk 8 above)
Comment (sveccham) [reply chemprop/scripts/check_and_install_cuik_molmaker.py:38]: `cuik-molmaker` only has Linux builds at the moment. I have added some checks so that the install script fails when running on windows or MacOS.

File: chemprop/cli/train.py (hunk 1 above)
Comment (sveccham) [reply chemprop/cli/train.py:987]: Sounds good. I can implement it in another PR. I have created an issue in the meanwhile.
https://github.com/chemprop/chemprop/issues/1256

File: chemprop/cli/common.py [hunk 11]
```diff
@@ -222,7 +234,29 @@ def process_common_args(args: Namespace) -> Namespace:
 
 
 def validate_common_args(args):
-    pass
+    # Validation that cuik-molmaker is installed if the flag is used
+    if args.use_cuikmolmaker_featurization and not is_cuikmolmaker_available():
+        raise ArgumentError(
+            argument=None,
+            message="cuik-molmaker is not installed. Please install it using `python chemprop/scripts/check_and_install_cuik_molmaker.py` before using the `--use-cuikmolmaker-featurization` flag.",
+        )
```
[the remark this answers is not in the document]
Comment (sveccham) [reply chemprop/cli/common.py:242]: double-checked and made the changes

File: chemprop/cli/utils/parsing.py (hunk 3 above)
Comment (sveccham) [reply chemprop/cli/utils/parsing.py]: I compute `X_d` and set `x_d` later on. I am doing this to ensure that the RDKit molecule object is created only once.

**Comment by sveccham** [comment]
> Thanks for you and your team's outstanding work on this PR! We're very excited to see chemprop scale up larger and larger!
> 
> A couple overall notes:
> 
> * Rather than changing the behavior of `_MoleculeDatapointMixin` to be Lazy by default, I think a `_LazyMoleculeDatapointMixin` would be better - keeps existing code working exactly the same way.
> * The names `BatchMoleculeDataset` and `BatchMolGraphFeaturizer` make sense but obscure the fact that they use `cuik-molmaker`. For the sake of clarity, I think they should be renamed something along the lines of `CuikMolMakerDataset` and `CuikMolmakerFeaturizer` or something in that direction
> 
> I've not yet looked at the notebook or tests - will review those after this initial round (or let another reviewer take a look, perhaps).

Thanks for your notes. I have added a `_LazyMoleculeDatapointMixin` and `LazyMoleculeDatapoint` object and made `CuikmolmakerDataset` (previously `BatchMoleculeDataset`) use those classes. I have also renamed the featurizer, dataset, and collate functions to reflect the fact that they use `cuik-molmaker`.

File: chemprop/scripts/check_and_install_cuik_molmaker.py (hunk 7 above)
Comment (sveccham) [reply chemprop/scripts/check_and_install_cuik_molmaker.py]: done

**Comment by sveccham** [comment]
@JacksonBurns , I have made the changes request and addressed all of your comments. Could you take another look? 
All the tests pass on my local ubuntu 3.11 and 3.12 environments. Could you trigger the CI again as well?

**Comment by sveccham** [comment]
@JacksonBurns , thanks for triggering CI. Is the `tests/integration/test_regression_MAB.py` test known to be flaky?

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
Thanks for making all these changes! Things are looking great now. Small thing here, and then let's continue this chat in https://github.com/sveccham/chemprop/pull/1

File: chemprop/scripts/check_and_install_cuik_molmaker.py (hunk 6 above)
Comment (sveccham) [reply chemprop/scripts/check_and_install_cuik_molmaker.py]: I doubled checked this. So if install `chemprop` from `pip`, should (1) `pip uninstall rdkit` and (2) `conda install conda-forge::rdkit`. Now, `cuik-molmaker` will work with this conda installed `rdkit`.

**Comment by sveccham** [comment]
@JacksonBurns , is autoapi generated running `check_and_install_cuik_molmaker.py`?
https://app.readthedocs.org/projects/chemprop/builds/28726110/#277045996--736

**Comment by sveccham** [comment]
> > @JacksonBurns , is autoapi generating running `check_and_install_cuik_molmaker.py` script? https://app.readthedocs.org/projects/chemprop/builds/28726110/#277045996--736
> 
> @JacksonBurns documentation build works now. If everything looks good to you, I would love an approval on this PR? In the meanwhile, would be useful to check CI again.

@JacksonBurns gentle reminder about this. Let me know if there is anything I can do to push this over the finish line.
